File: razer-sdk/src/test/java/net/booone/razersdk/sdk/ProgressiveRzKeySelectorTest.java
package net.booone.razersdk.sdk;

import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class ProgressiveRzKeySelectorTest {

    @Test
    void testJoinSets() {
        //given
        Set<RzKey> set1 = Set.of(RzKey.RZKEY_Q, RzKey.RZKEY_A, RzKey.RZKEY_Z);
        Set<RzKey> set2 = Set.of(RzKey.RZKEY_Q, RzKey.RZKEY_R);

        //when
        Set<RzKey> actual = joinSets(set1, set2);

        //then
        assertEquals(Set.of(RzKey.RZKEY_Q, RzKey.RZKEY_A, RzKey.RZKEY_Z, RzKey.RZKEY_R), actual);
    }

    @Test
    void testIllegalLengthException() {
        //given
        List<Set<RzKey>> testSet = List.of(Set.of(RzKey.RZKEY_ENTER));

        //when
        assertThrows(IllegalArgumentException.class, () -> new ProgressiveRzKeySelector(testSet, -1));
        assertThrows(IllegalArgumentException.class, () -> new ProgressiveRzKeySelector(testSet, 0));
        assertThrows(IllegalArgumentException.class, () -> new ProgressiveRzKeySelector(testSet, 2));
    }

    @Test
    void testIllegalPartsException() {
        //when
        assertThrows(IllegalArgumentException.class, () -> new ProgressiveRzKeySelector(null));
        assertThrows(IllegalArgumentException.class, () -> new ProgressiveRzKeySelector(Collections.emptyList()));
    }

    @Test
    void testLength1Progression() {
        //given
        Set<RzKey> set1 = Set.of(RzKey.RZKEY_Q, RzKey.RZKEY_A, RzKey.RZKEY_Z);
        RzKey[] array2 = new RzKey[]{RzKey.RZKEY_W, RzKey.RZKEY_S};
        RzKey rzKey3 = RzKey.RZKEY_E;

        //when
        ProgressiveRzKeySelector progressiveRzKeySelector = new ProgressiveRzKeySelector.Builder()
                .addPart(set1)
                .addPart(array2)
                .addPart(rzKey3)
                .build();

        //then
        assertEquals(set1, progressiveRzKeySelector.getNextPart());
        assertEquals(Arrays.stream(array2).collect(Collectors.toSet()), progressiveRzKeySelector.getNextPart());
        assertEquals(Set.of(rzKey3), progressiveRzKeySelector.getNextPart());

        assertEquals(set1, progressiveRzKeySelector.getNextPart());
        assertEquals(Arrays.stream(array2).collect(Collectors.toSet()), progressiveRzKeySelector.getNextPart());
        assertEquals(Set.of(rzKey3), progressiveRzKeySelector.getNextPart());
    }

    @Test
    void testLength2Progression() {
        //given
        Set<RzKey> set1 = Set.of(RzKey.RZKEY_Q, RzKey.RZKEY_A, RzKey.RZKEY_Z);
        Set<RzKey> set2 = Set.of(RzKey.RZKEY_W, RzKey.RZKEY_S);
        Set<RzKey> set3 = Set.of(RzKey.RZKEY_E);

        //when
        ProgressiveRzKeySelector progressiveRzKeySelector = new ProgressiveRzKeySelector.Builder()
                .addPart(set1)
                .addPart(set2)
                .addPart(set3)
                .withLength(2)
                .build();

        //then
        assertEquals(set1, progressiveRzKeySelector.getNextPart());
        assertEquals(joinSets(set1, set2), progressiveRzKeySelector.getNextPart());
        assertEquals(joinSets(set2, set3), progressiveRzKeySelector.getNextPart());
        assertEquals(set3, progressiveRzKeySelector.getNextPart());

        assertEquals(set1, progressiveRzKeySelector.getNextPart());
        assertEquals(joinSets(set1, set2), progressiveRzKeySelector.getNextPart());
        assertEquals(joinSets(set2, set3), progressiveRzKeySelector.getNextPart());
        assertEquals(set3, progressiveRzKeySelector.getNextPart());
    }

    @Test
    void testLength3Progression() {
        //given
        Set<RzKey> set1 = Set.of(RzKey.RZKEY_Q, RzKey.RZKEY_A, RzKey.RZKEY_Z);
        Set<RzKey> set2 = Set.of(RzKey.RZKEY_W, RzKey.RZKEY_S);
        Set<RzKey> set3 = Set.of(RzKey.RZKEY_E);

        //when
        ProgressiveRzKeySelector progressiveRzKeySelector = new ProgressiveRzKeySelector(List.of(set1, set2, set3), 3);

        //then
        assertEquals(set1, progressiveRzKeySelector.getNextPart());
        assertEquals(joinSets(set1, set2), progressiveRzKeySelector.getNextPart());
        assertEquals(joinSets(set1, set2, set3), progressiveRzKeySelector.getNextPart());
        assertEquals(joinSets(set2, set3), progressiveRzKeySelector.getNextPart());
        assertEquals(set3, progressiveRzKeySelector.getNextPart());

        assertEquals(set1, progressiveRzKeySelector.getNextPart());
        assertEquals(joinSets(set1, set2), progressiveRzKeySelector.getNextPart());
        assertEquals(joinSets(set1, set2, set3), progressiveRzKeySelector.getNextPart());
        assertEquals(joinSets(set2, set3), progressiveRzKeySelector.getNextPart());
        assertEquals(set3, progressiveRzKeySelector.getNextPart());
    }

    @SafeVarargs
    static Set<RzKey> joinSets(Set<RzKey>... sets) {
        return Stream.of(sets).flatMap(Set::stream).collect(Collectors.toSet());
    }
}

File: test/razer-sdk/sdk/ProgressiveRzKeySelector.test.ts
import {describe, expect, test} from 'vitest';
import {
    ProgressiveRzKeySelector,
    ProgressiveRzKeySelectorBuilder
} from '../../../src/razer-sdk/sdk/ProgressiveRzKeySelector.js';
import {RzKey} from '../../../src/razer-sdk/sdk/RzKey.js';

function joinSets(...sets: Set<RzKey>[]): Set<RzKey> {
    const result = new Set<RzKey>();
    for (const set of sets) {
        for (const key of set) {
            result.add(key);
        }
    }
    return result;
}

describe('ProgressiveRzKeySelectorTest', () => {
    test('testJoinSets', () => {
        //given
        const set1 = new Set([RzKey.RZKEY_Q, RzKey.RZKEY_A, RzKey.RZKEY_Z]);
        const set2 = new Set([RzKey.RZKEY_Q, RzKey.RZKEY_R]);

        //when
        const actual = joinSets(set1, set2);

        //then
        expect(actual).toEqual(new Set([RzKey.RZKEY_Q, RzKey.RZKEY_A, RzKey.RZKEY_Z, RzKey.RZKEY_R]));
    });

    test('testIllegalLengthException', () => {
        //given
        const testSet = [new Set([RzKey.RZKEY_ENTER])];

        //when
        expect(() => new ProgressiveRzKeySelector(testSet, -1)).toThrow();
        expect(() => new ProgressiveRzKeySelector(testSet, 0)).toThrow();
        expect(() => new ProgressiveRzKeySelector(testSet, 2)).toThrow();
    });

    test('testIllegalPartsException', () => {
        //when
        expect(() => new ProgressiveRzKeySelector(null as any)).toThrow();
        expect(() => new ProgressiveRzKeySelector([])).toThrow();
    });

    test('testLength1Progression', () => {
        //given
        const set1 = new Set([RzKey.RZKEY_Q, RzKey.RZKEY_A, RzKey.RZKEY_Z]);
        const array2 = [RzKey.RZKEY_W, RzKey.RZKEY_S];
        const rzKey3 = RzKey.RZKEY_E;

        //when
        const progressiveRzKeySelector = new ProgressiveRzKeySelectorBuilder()
            .addPart(set1)
            .addPart(array2)
            .addPartFromKeys(rzKey3)
            .build();

        //then
        expect(progressiveRzKeySelector.getNextPart()).toEqual(set1);
        expect(progressiveRzKeySelector.getNextPart()).toEqual(new Set(array2));
        expect(progressiveRzKeySelector.getNextPart()).toEqual(new Set([rzKey3]));

        expect(progressiveRzKeySelector.getNextPart()).toEqual(set1);
        expect(progressiveRzKeySelector.getNextPart()).toEqual(new Set(array2));
        expect(progressiveRzKeySelector.getNextPart()).toEqual(new Set([rzKey3]));
    });

    test('testLength2Progression', () => {
        //given
        const set1 = new Set([RzKey.RZKEY_Q, RzKey.RZKEY_A, RzKey.RZKEY_Z]);
        const set2 = new Set([RzKey.RZKEY_W, RzKey.RZKEY_S]);
        const set3 = new Set([RzKey.RZKEY_E]);

        //when
        const progressiveRzKeySelector = new ProgressiveRzKeySelectorBuilder()
            .addPart(set1)
            .addPart(set2)
            .addPart(set3)
            .withLength(2)
            .build();

        //then
        expect(progressiveRzKeySelector.getNextPart()).toEqual(set1);
        expect(progressiveRzKeySelector.getNextPart()).toEqual(joinSets(set1, set2));
        expect(progressiveRzKeySelector.getNextPart()).toEqual(joinSets(set2, set3));
        expect(progressiveRzKeySelector.getNextPart()).toEqual(set3);

        expect(progressiveRzKeySelector.getNextPart()).toEqual(set1);
        expect(progressiveRzKeySelector.getNextPart()).toEqual(joinSets(set1, set2));
        expect(progressiveRzKeySelector.getNextPart()).toEqual(joinSets(set2, set3));
        expect(progressiveRzKeySelector.getNextPart()).toEqual(set3);
    });

    test('testLength3Progression', () => {
        //given
        const set1 = new Set([RzKey.RZKEY_Q, RzKey.RZKEY_A, RzKey.RZKEY_Z]);
        const set2 = new Set([RzKey.RZKEY_W, RzKey.RZKEY_S]);
        const set3 = new Set([RzKey.RZKEY_E]);

        //when
        const progressiveRzKeySelector = new ProgressiveRzKeySelector([set1, set2, set3], 3);

        //then
        expect(progressiveRzKeySelector.getNextPart()).toEqual(set1);
        expect(progressiveRzKeySelector.getNextPart()).toEqual(joinSets(set1, set2));
        expect(progressiveRzKeySelector.getNextPart()).toEqual(joinSets(set1, set2, set3));
        expect(progressiveRzKeySelector.getNextPart()).toEqual(joinSets(set2, set3));
        expect(progressiveRzKeySelector.getNextPart()).toEqual(set3);

        expect(progressiveRzKeySelector.getNextPart()).toEqual(set1);
        expect(progressiveRzKeySelector.getNextPart()).toEqual(joinSets(set1, set2));
        expect(progressiveRzKeySelector.getNextPart()).toEqual(joinSets(set1, set2, set3));
        expect(progressiveRzKeySelector.getNextPart()).toEqual(joinSets(set2, set3));
        expect(progressiveRzKeySelector.getNextPart()).toEqual(set3);
    });
});
